File: 백준/Gold/15686. 치킨 배달/치킨 배달.swift
import Foundation

let input = readLine()!.split(separator: " ")
    .map({ Int($0)! })

let n = input[0]
let m = input[1]

var houses: [(r: Int, c: Int)] = []
var chickens: [(r: Int, c: Int)] = []

for r in 0..<n {
    let row = readLine()!.split(separator: " ").map({ Int($0)! })
    for c in 0..<n {
        if row[c] == 1 { houses.append((r, c)) }
        if row[c] == 2 { chickens.append((r, c)) }
    }
}

var distTable = Array(repeating: [Int](), count: houses.count)

for (hIdx, house) in houses.enumerated() {
    for chicken in chickens {
        let dist = abs(house.r - chicken.r) + abs(house.c - chicken.c)
        distTable[hIdx].append(dist)
    }
}

var result = Int.max
var visited = Array(repeating: false, count: chickens.count)

func dfs(index: Int, count: Int) {
    if count == m {
        var currentCityDist = 0
        
        for hIdx in 0..<houses.count {
            var minHouseDist = Int.max
            
            for cIdx in 0..<chickens.count {
                if visited[cIdx] {
                    let d = distTable[hIdx][cIdx]
                    if d < minHouseDist {
                        minHouseDist = d
                    }
                }
            }
            currentCityDist += minHouseDist
            
            if currentCityDist >= result { return }
        }
        
        if currentCityDist < result {
            result = currentCityDist
        }
        return
    }
    
    for i in index..<chickens.count {
        visited[i] = true
        dfs(index: i + 1, count: count + 1)
        visited[i] = false
    }
}

dfs(index: 0, count: 0)
print(result)

File: 백준/Gold/15686. 치킨 배달/치킨 배달.cc
#include <bits/stdc++.h>
using namespace std;

int n, m;
int result = INT_MAX;

struct Point {
    int r, c;
};

vector<Point> houses;
vector<Point> chickens;

// distTable[집인덱스][치킨집인덱스] = 거리
vector<vector<int>> distTable;
vector<bool> visited;

void dfs(int index, int count) {
    if (count == m) {
        int currentCityDist = 0;

        for (int hIdx = 0; hIdx < houses.size(); ++hIdx) {
            int minHouseDist = INT_MAX;

            // 선택된 치킨집들(visited가 true인 곳) 중에서 가장 가까운 거리 찾기
            for (int cIdx = 0; cIdx < chickens.size(); ++cIdx) {
                if (visited[cIdx]) {
                    int d = distTable[hIdx][cIdx];
                    if (d < minHouseDist) {
                        minHouseDist = d;
                    }
                }
            }
            currentCityDist += minHouseDist;

            if (currentCityDist >= result) return;
        }

        if (currentCityDist < result) {
            result = currentCityDist;
        }
        return;
    }
    
    for (int i = index; i < chickens.size(); ++i) {
        visited[i] = true;
        dfs(i + 1, count + 1);
        visited[i] = false;
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    if (!(cin >> n >> m)) return 0;

    for (int r = 0; r < n; ++r) {
        for (int c = 0; c < n; ++c) {
            int value;
            cin >> value;
            if (value == 1) {
                houses.push_back({r, c});
            } else if (value == 2) {
                chickens.push_back({r, c});
            }
        }
    }

    distTable.resize(houses.size());

    for (int hIdx = 0; hIdx < houses.size(); ++hIdx) {
        for (int cIdx = 0; cIdx < chickens.size(); ++cIdx) {
            int dist = abs(houses[hIdx].r - chickens[cIdx].r) +
                       abs(houses[hIdx].c - chickens[cIdx].c);
            distTable[hIdx].push_back(dist);
        }
    }

    visited.resize(chickens.size(), false);

    dfs(0, 0);
    cout << result << "\n";
    return 0;
}